### PR TITLE
Implement JSONiq functions

### PR DESCRIPTION
**Pull request by riyafa**
…rray and null

### CODE REVIEW

File: vxquery-core/src/main/java/org/apache/vxquery/runtime/functions/misc/FnDataScalarEvaluatorFactory.java [hunk 1]
```diff
@@ -71,10 +71,12 @@ protected void evaluate(TaggedValuePointable[] args, IPointable result) throws S
                     int seqLen = seq.getEntryCount();
                     for (int j = 0; j < seqLen; ++j) {
                         seq.getEntry(j, p);
+                        check(p);
                         ah.atomize(p, ppool, tempTVP);
```
Comment (prestoncarman) [inline vxquery-core/src/main/java/org/apache/vxquery/runtime/functions/misc/FnDataScalarEvaluatorFactory.java:74]: Move the check to atomize helper class. This way any user of the atomize helper class will get the new error message.

File: vxquery-core/src/main/java/org/apache/vxquery/runtime/functions/cast/AbstractCastToOperation.java [hunk 2]
```diff
@@ -226,4 +226,8 @@ public void convertNormalizedString(UTF8StringPointable stringp, DataOutput dOut
     public void convertToken(UTF8StringPointable stringp, DataOutput dOut) throws SystemException, IOException {
         throw new SystemException(ErrorCode.XPTY0004);
     }
+
+    public void convertNull(DataOutput dOut) throws SystemException, IOException {
```
Comment (prestoncarman) [inline vxquery-core/src/main/java/org/apache/vxquery/runtime/functions/cast/AbstractCastToOperation.java:230]: Also update CastableAsStringOperation for the null case.

Comment (prestoncarman) [inline vxquery-core/src/main/java/org/apache/vxquery/runtime/functions/cast/AbstractCastToOperation.java:230]: Also add a test case for this option.

Comment (riyafa) [inline vxquery-core/src/main/java/org/apache/vxquery/runtime/functions/cast/AbstractCastToOperation.java:230]: I think I have already added a testcase for null with fn:string in file fn_string1.xq. What do you mean by "this options"? Does anything change if I add null to CastableAsStringOperation that needs to be tested?

Comment (prestoncarman) [inline vxquery-core/src/main/java/org/apache/vxquery/runtime/functions/cast/AbstractCastToOperation.java:230]: Here is an example query: 

xs:string("test") castable as js:null

Does this help to explain? Add a test query with the above syntax and update CastableAsStringOperation similar to CastAsStringOperation.

Comment (prestoncarman) [inline vxquery-core/src/main/java/org/apache/vxquery/runtime/functions/cast/AbstractCastToOperation.java:230]: I see you update the code, now we just need a query to test it.

Comment (prestoncarman) [inline vxquery-core/src/main/java/org/apache/vxquery/runtime/functions/cast/AbstractCastToOperation.java:230]: Also, add a direct test for casting:
xs:string("test") cast as js:null

Change the query as necessary for your test.

**Comment by prestoncarman** [comment]
One last comment: Why comment out one of the index tests?
Ready to squash commits. If you need to start a new branch and do a squash merge to get a single commit. Remember to make a nice commit message like an e-mail: first line is a title (<50 chars) and line break then the body.

**Comment by riyafa** [comment]
When I comment that test all the tests pass. I will uncomment it and create a pull request with only one commit.

**Comment by prestoncarman** [comment]
Go ahead and leave it commented out if the tests pass.

File: vxquery-core/src/main/java/org/apache/vxquery/runtime/functions/cast/CastScalarEvaluatorFactory.java [hunk 3]
```diff
@@ -56,9 +56,9 @@ protected void evaluate(TaggedValuePointable tvp, IPointable result) throws Syst
                 int tid = tvp.getTag();
                 try {
                     switch (tid) {
-                    /**
-                     * Primitive Datatypes (Alphabetical)
-                     */
+                        /**
+                         * Primitive Datatypes (Alphabetical)
+                         */
```
Comment (westmann) [inline vxquery-core/src/main/java/org/apache/vxquery/runtime/functions/cast/CastScalarEvaluatorFactory.java:61]: Please remove the tabs.

**Comment by westmann** [comment]
Looks good to me as well.

**Comment by prestoncarman** [comment]
Merged the squashed version of this change.

**Pull request closed without merging**